src/Waargonaut/Decode.hs

```diff
@@ -525,6 +527,49 @@ foldCursor nom f s elemD curs = DecodeResult . ReaderT $ \p ->
     (DI.Decoder' $ runDecoder elemD p)
     curs
 
+-- | Helper function for "pattern matching" on a decoded value to some Haskell
+-- value. The 'Text' argument is used in the error message should this decoder
+-- fail. Normally it would simply be the name of the type you are writing the
+-- decoder for.
+--
+-- This is useful for decoding sum types, such as:
+--
+-- @
+-- data MyEnum
+--   = A
+--   | B
+--   | C
+--
+-- decodeMyEnum :: Monad f => Decoder f MyEnum
+-- decodeMyEnum = D.oneOf D.text "MyEnum"
+--   [ ("a", A)
+--   , ("b", B)
+--   , ("c", C)
+--   ]
+--
+-- decodeMyEnumFromInt :: Monad f => Decoder f MyEnum
+-- decodeMyEnumFromInt = D.oneOf D.int "MyEnum"
+--   [ (1, A)
+--   , (2, B)
+--   , (3, C)
+--   ]
+-- @
+--
+oneOf
+  :: ( Foldable g
+     , Functor g
+     , Monad f
+     , Eq a
+     )
+  => Decoder f a
+  -> Text
+  -> g (a, b)
+  -> Decoder f b
+oneOf d l = foldr (<!>) err
+  . fmap (\(a,b) -> d >>= \t -> if t == a then pure b else err)
```

Comment:
This can be rewritten to only use foldr, and thereby drop the `Functor` constraint. I'd find that preferable.

Reply:
Done :+1: